feat(ai-gateway): Add new getting started guide for proxying LLM traffic (#5637)
Kong/developer.konghq.com: app/_how-tos/ai-gateway/get-started-with-ai-gateway.md

title: Get started with {{site.ai_gateway}}
content_type: how_to
permalink: /ai-gateway/get-started/
description: Learn how to proxy LLM traffic with {{site.ai_gateway}} entities in {{site.konnect_product_name}}
products:
  - ai-gateway

works_on:
  - konnect

entities:
  - ai-provider
  - ai-model

tags:
  - get-started
  - ai

tldr:
  q: How do I proxy LLM traffic with {{site.ai_gateway}} entities?
  a: |
    {{site.ai_gateway}} provides first-class entities for managing LLM providers and models in {{site.konnect_product_name}}.

>     Create a Provider entity to connect to an LLM service like OpenAI, then create Model entities

Reviewer: ```suggestion
    Create an AI Provider entity to connect to an LLM service like OpenAI, then create AI Model entities
```

    to specify which models are available for requests.

>     This tutorial shows you how to set up a Provider and Model in {{site.konnect_product_name}} using the {{site.konnect_product_name}} API.

Reviewer: ```suggestion
    This tutorial shows you how to set up an AI Provider and AI Model in {{site.konnect_product_name}} using the {{site.konnect_product_name}} API.
```

tools:
  - konnect-api

prereqs:
  inline:
    - title: OpenAI credentials
      content: |
        This tutorial uses OpenAI as the LLM provider. You'll need to [create an OpenAI account](https://auth.openai.com/create-account)
        and [get an API key](https://platform.openai.com/api-keys). Save your API key for the next steps:

        ```sh
        export OPENAI_API_KEY='<api-key>'
        ```
cleanup:
  inline:
    - title: Clean up {{site.ai_gateway}} resources
      include_content: cleanup/products/ai-gateway

min_version:
  ai-gateway: '2.0'

---

> ## Create a Provider entity

Reviewer: ```suggestion
## Create an AI Provider entity
```

> Create a [Provider](/ai-gateway/entities/ai-provider/) entity to define your LLM service and store authentication credentials:

Reviewer: ```suggestion
Create an [AI Provider](/ai-gateway/entities/ai-provider/) entity to define your LLM Service and store authentication credentials:
```

<!-- vale off -->
{% konnect_api_request %}
url: /v1/ai-gateways/$AI_GATEWAY_ID/providers
status_code: 201
method: POST
headers:
  - 'Content-Type: application/json'
  - 'Accept: application/json, application/problem+json'
body:
  display_name: OpenAI Production
  name: my-openai-account
  type: openai
  config:
    auth:
      type: basic
      headers:
        - name: Authorization
          value: Bearer $OPENAI_API_KEY
{% endkonnect_api_request %}
<!-- vale on -->

> Save the Provider ID from the response for cleanup:

Reviewer: ```suggestion
Save the AI Provider ID from the response:
```

```bash
export PROVIDER_ID=ai-provider-id
```

> ## Create a Model entity

Reviewer: ```suggestion
## Create an AI Model entity
```

> Create a [Model](/ai-gateway/entities/ai-model/) entity to specify which LLM models are available and declare their capabilities. Each capability generates a route on the service:

Reviewer: ```suggestion
Create an [AI Model](/ai-gateway/entities/ai-model/) entity to specify which LLM models are available and declare their capabilities. Each capability generates a Route on the Service:
```

<!-- vale off -->
{% konnect_api_request %}
url: /v1/ai-gateways/$AI_GATEWAY_ID/models
status_code: 201
method: POST
headers:
  - 'Content-Type: application/json'
  - 'Accept: application/json, application/problem+json'
body:
  display_name: GPT-4o
  name: my-gpt-4o
  type: model
  enabled: true
  capabilities:
    - generate
  formats:
    - type: openai
  config:

>     route:

Reviewer: ```suggestion
    model: 
      alias: gtp-4o
    route:
```

I got an error saying config.model was missing. According to the spec it's a required field, but neither of its properties are required, so I'm not sure what the point is.

Author: This is weird, I am checking with the koko team, it shouldn't be required

      paths:
        - /v1
  target_models:
    - name: gpt-4o

>       provider: openai-provider

Reviewer: ```suggestion
      provider: my-openai-account
      config:
        type: openai
```

I got an error about target_models.config too

{% endkonnect_api_request %}
<!-- vale on -->

> Save the Model ID from the response for cleanup:

Reviewer: ```suggestion
Save the AI Model ID from the response:
```

```bash
export MODEL_ID=ai-model-id
```

{:.info}

> > The `generate` capability creates a `/chat/completions` route. The `paths: ["/v1"]` setting defines the base path, so the final route becomes `/v1/chat/completions`.

Reviewer: ```suggestion
> The `generate` capability creates a `/chat/completions` Route. The `paths: ["/v1"]` setting defines the base path, so the final Route becomes `/v1/chat/completions`.
```

## Validate

Send a chat request to verify your setup:

<!-- vale off -->
{% validation request-check %}
url: /v1/chat/completions
status_code: 200
method: POST
headers:
    - 'Accept: application/json'
    - 'Content-Type: application/json'
    - 'Authorization: Bearer $OPENAI_API_KEY'
body:
  model: my-gpt-4o
  messages:
  - role: "user"
    content: "Say this is a test!"

> {% endvalidation %}

Reviewer: I'm getting this error:
> curl: (22) The requested URL returned error: 500
> {"error":{"message":"failed to get balancer instance"}}

title: Get started with {{site.ai_gateway}}
content_type: how_to
permalink: /ai-gateway/get-started/
description: Learn how to proxy LLM traffic with {{site.ai_gateway}} entities in {{site.konnect_product_name}}
products:
  - ai-gateway

works_on:
  - konnect

entities:
  - ai-provider
  - ai-model

tags:
  - get-started
  - ai

tldr:
  q: How do I proxy LLM traffic with {{site.ai_gateway}} entities?
  a: |
    {{site.ai_gateway}} provides first-class entities for managing LLM providers and models in {{site.konnect_product_name}}.
    Create an [AI Provider](/ai-gateway/entities/ai-provider/) entity to connect and authenticate to an LLM service like OpenAI, then create an [AI
    Model](/ai-gateway/entities/ai-model/) entity to specify which model is available for requests.

>     This tutorial shows you how to set up an AI Provider and AI Model for OpenAI in {{site.konnect_product_name}} using the {{site.konnect_product_name}} API and how to proxy your first request to OpenAI.

Reviewer: ```suggestion
    This tutorial shows you how to set up an AI Provider and AI Model for OpenAI in {{site.konnect_product_name}} using the {{site.konnect_short_name}} API and how to proxy your first request to OpenAI.
```

## Create an AI Model entity

Create an [AI Model](/ai-gateway/entities/ai-model/) entity to declare which upstream models are available, configure how client requests are routed, and specify which AI Provider to use:

<!-- vale off -->
{% konnect_api_request %}
url: /v1/ai-gateways/$AI_GATEWAY_ID/models
status_code: 201
method: POST
headers:
  - 'Content-Type: application/json'
  - 'Accept: application/json, application/problem+json'
body:
  display_name: my-gpt-4o
  name: my-gpt-4o
  type: model
  formats:
    - type: openai
  config:
    route:
      paths:
        - /v1
    model: {}
    logging:
      payloads: false
      statistics: true
  targets:
    - name: gpt-4o
      provider: generic-openai
      config:
        type: openai
  policies: []
  capabilities:
    - generate
{% endkonnect_api_request %}
<!-- vale on -->

In this example, we're setting up the AI Model with:

* `type: model`: Specifies this is a synchronous model for request/response workloads.
* `name: my-gpt-4o`: A unique identifier for this model.
* `formats: [type: openai]`: Declares that this model accepts requests in OpenAI-compatible format.

> * `config.route.paths: [/v1]`: Configures the custom base path where this model's routes will be accessible. Clients will send requests to paths that combine this base path with capability-specific routes.

Reviewer: ```suggestion
* `config.route.paths: [/v1]`: Configures the custom base path where this model's Routes will be accessible. Clients will send requests to paths that combine this base path with capability-specific Routes.
```

* `capabilities: [generate]`: Enables the text generation capability. The `generate` capability creates a `/chat/completions` endpoint, so combined with your base path, clients send chat requests to `/v1/chat/completions`.

> * `targets`: Specifies which upstream AI Provider model to route requests to. Here, `provider: generic-openai` references the AI Provider you created earlier, and `name: gpt-4o` specifies which OpenAI model to call upstream.

Reviewer: ```suggestion
* `targets`: Specifies which upstream AI Provider model to route requests to. Here, `provider: generic-openai` references the AI Provider we created earlier, and `name: gpt-4o` specifies which OpenAI model to call upstream.
```

* `config.logging`: Configures what gets logged. With `statistics: true`, usage metrics (tokens, latency, cost) are logged for monitoring and billing. With `payloads: false`, full request/response bodies are not logged for privacy.

## Validate

Send a chat request to verify your setup:

<!-- vale off -->
{% validation request-check %}
url: /v1/chat/completions
status_code: 200
method: POST
headers:
    - 'Accept: application/json'
    - 'Content-Type: application/json'

>     - 'Authorization: Bearer $OPENAI_API_KEY'

Reviewer: Do we need this? In the first step it says:
> `config.auth`: Stores your OpenAI API key. {{site.ai_gateway}} securely manages this credential and injects it into upstream requests automatically, eliminating the need for clients to pass API keys.
```suggestion
```
